keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1187 @@
+requests from the same origin are always directed back to the same leader pod.
+
+### API Specification
+
+The section describes the specification of proposed API. The Go types of the CRD
+are defined as follows:
+
+```go
+// VolumeSnapshotDelta is a specification for a VolumeSnapshotDelta resource
+type VolumeSnapshotDelta struct {
+  metav1.TypeMeta   `json:",inline"`
+
+  // +optional
+  metav1.ObjectMeta `json:"metadata,omitempty"`
+
+  Spec   VolumeSnapshotDeltaSpec   `json:"spec"`
+
+  // +optional
+  Status VolumeSnapshotDeltaStatus `json:"status,omitempty"`
+}
+
+// VolumeSnapshotDeltaSpec is the spec for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaSpec struct {
+  BaseVolumeSnapshotName   string `json:"baseVolumeSnapshotName,omitempty"` // name of the base VolumeSnapshot; optional
+  TargetVolumeSnapshotName string `json:"targetVolumeSnapshotName"`        // name of the target VolumeSnapshot; required
+  Mode                     string `json:"mode,omitempty"`         // default to "block"
+}
+
+// VolumeSnapshotDeltaStatus is the status for a VolumeSnapshotDelta resource
+type VolumeSnapshotDeltaStatus struct {
+  Error        string `json:"error,omitempty"`
+  State        string `json:"state"`
+  CallbackURL  string `json:"callbackURL"`
+}
+
+// VolumeSnapshotDeltaList is a list of VolumeSnapshotDelta resources
+type VolumeSnapshotDeltaList struct {
+  metav1.TypeMeta          `json:",inline"`
+  metav1.ListMeta          `json:"metadata"`
+  Items []VolumeSnapshotDelta `json:"items"`
+}
+```
+
+The corresponding GRPC service definition is as follows:
+
+```grpc
+service VolumeSnapshotDelta {
+  rpc GetVolumeSnapshotDelta(VolumeSnapshotDeltaRequest)
+    returns (VolumeSnapshotDeltaResponse) {}
+}
+
+type VolumeSnapshotDeltaRequest struct {
+    // If SnapshotBase is not specified, return all used blocks.
+    SnapshotBase       string         // Snapshot handle, optional.
+    SnapshotTarget     string         // Snapshot handle, required.
+    Mode               string         // Volume mode, default "Block"
+    StartOffset        string         // Logical offset from beginning of disk/volume.
+                                      // Use string instead of uint64 to give vendor
+                                      // the flexibility of implementing it either
+                                      // string "token" or a number.
+    MaxEntries         uint64         // Maximum number of entries in the response
+}
+
+type VolumeSnapshotDeltaResponse struct {
+    ChangeBlockList   []ChangedBlock  // array of ChangedBlock (for "Block" mode)
+    NextOffset        string          // StartOffset of the next “page”.
+    VolumeSize        uint64          // size of volume in bytes
+    Timeout           uint64          // Time when the result of this differential snapshot is
+                                      // available in the backend storage.  Seconds since epoch.
```

Comment:
In what timezone is the time? I'd suggest to avoid absolute time, but relative timeout in seconds since the response was sent. And what will the backup SW do with the timeout anyway?

Reply:
The original intent is that some providers return a temporary `DataToken` which can be used to retrieve the actual data block, with `Timeout` being the token's TTL. I think we should introduce a new message type called `DataToken` to include the `token` and the `ttl` fields, to avoid confusion with the pagination token. WDYT?

Reply:
I can't tell without knowing how the actual protocol to read the block will look like. And it's intentionally missing in this KEP.